### Reduction operation: the sum of the numbers in the range [0, a)

In [6]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.28615713119506836 seconds
269 ms ± 210 µs per loop (mean ± std. dev. of 2 runs, 1 loop each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



### 3.2. Python HPC: lists y Numpy y Numba con Jupyter notebook

In [15]:
# A) Crear una lista de Python con la función range que contenga 10^6 elementos y calcular el tiempo que tarda en sumar todos los elementos de dos maneras diferentes:
lista = list(range(1_000_000)) # Creamos la lsita
import time # Importamos la librería time

# Usando un bucle for:

# Definimos la función para sumar
def sumar(data):
    total = 0
    for num in lista:
        total += num
    return total

# Medimos el tiempo con time
start = time.time()
total = sumar(lista)
end = time.time()
print(f"Resultado de la suma: {total}")
print(f"\nTiempo con time usando un bucle for: {end - start:.6f} segundos")

# Medimos el tiempo con %timeit
print("Tiempo con %timeit usando un bucle for:")
%timeit sumar(lista)

# Usando la función sum:

start = time.time()
total = sum(lista)
end = time.time()
print(f"\nTiempo con time usando la función sum: {end - start:.6} segundos")
print("Tiempo con %timeit usando la función sum:")
%timeit sum(lista)

Resultado de la suma: 499999500000

Tiempo con time usando un bucle for: 0.206934 segundos
Tiempo con %timeit usando un bucle for:
205 ms ± 5.22 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

Tiempo con time usando la función sum: 0.0314982 segundos
Tiempo con %timeit usando la función sum:
30.1 ms ± 496 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [16]:
# B) Pasar la lista creada a un array de Numpy. Calcular el tiempo en sumar todos los elementos del array:
import numpy as np
import time

# Creamos el array
array = np.array(lista)

# Usando un bucle for:

# Medimos el tiempo con time
start = time.time()
total = sumar(array)
end = time.time()

print(f"Resultado de la suma: {total}")
print(f"\nTiempo con time usando un bucle for: {end - start:.6f} segundos")

# Medimos el tiempo con %timeit
print("Tiempo con %timeit usando un bucle for:")
%timeit sumar(array)

# Usando la función sum del paquete Numpy.
start = time.time()
total = np.sum(array)
end = time.time()
print(f"\nTiempo usando np.sum: {end - start:.6f} segundos")
print("Tiempo con %timeit usando np.sum:")
%timeit np.sum(array)

Resultado de la suma: 499999500000

Tiempo con time usando un bucle for: 0.220886 segundos
Tiempo con %timeit usando un bucle for:
206 ms ± 3.24 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

Tiempo usando np.sum: 0.002308 segundos
Tiempo con %timeit usando np.sum:
1.38 ms ± 7.75 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


C) Explicación de los resultados:

El código original utiliza un bucle for para calcular la suma de un rango de números, obteniendo un tiempo de ejecución de aproximadamente 0.286 segundos con el método time. Al medir con %timeit, el tiempo promedio es de 269 ms ± 210 µs por loop, lo cual refleja la sobrecarga significativa del intérprete de Python al manejar iteraciones grandes. Este método es funcional, pero relativamente lento al operar con grandes cantidades de datos.

Cuando utilizamos listas, observamos una clara diferencia entre dos métodos: el bucle for, que tiene un tiempo de ejecución de aproximadamente 0.207 segundos con time y un promedio de 205 ms ± 5.22 ms por loop con %timeit, es más lento debido a la estructura dinámica de las listas. Por otro lado, la función sum reduce el tiempo de ejecución significativamente, a aproximadamente 0.031 segundos con time y 30.1 ms ± 496 µs por loop con %timeit. Esto ocurre porque sum utiliza optimizaciones internas que mejoran su eficiencia.

Al utilizar arrays de NumPy, el tiempo de ejecución varía notablemente según el método empleado. El bucle for, con un tiempo de 0.221 segundos con time y un promedio de 206 ms ± 3.24 ms por loop con %timeit, es el más lento, ya que no aprovecha las optimizaciones de NumPy y sigue dependiendo del intérprete de Python. Sin embargo, al usar la función optimizada np.sum, el tiempo de ejecución se reduce drásticamente a 0.002 segundos con time y 1.38 ms ± 7.75 µs por loop con %timeit. Esto se debe a que NumPy implementa operaciones vectorizadas en C, eliminando la sobrecarga de Python.

En conclusión, el código original y los métodos basados en bucles son intuitivos, pero significativamente más lentos para manejar grandes cantidades de datos. Las funciones optimizadas como sum y, especialmente, np.sum, son mucho más rápidas debido a su capacidad para aprovechar optimizaciones internas y operaciones vectorizadas.